Database Per-Range Lock Documentation (#11986)
apple/foundationdb: documentation/sphinx/source/rangelock.rst

###################################

> FoundationDB Per-Range Lock (WiP)

Reviewer: KeyRange or Range locks.
```suggestion
FoundationDB Range Locks (WiP)
```

###################################

| Author: Zhe Wang
| Reviewer: Jingyu Zhou
| Audience: FDB developers, SREs and expert users.

Overview
========

> Per-Range Lock is a feature that allows to shut down write traffic to a specific range of keys in FoundationDB (FDB).

Reviewer: Let not use "shut down".

```suggestion
KeyRange Lock is a feature that blocks write traffic to a specific key range in FoundationDB (FDB).
```

Author: Yes. "Block" is more concise to users

The locked range must be within the user key space, aka ``"" ~ \xff``.
If a user/app grabs a lock on a range, other user/app can read the range but cannot write to the range. Essentially, this lock type is
a readLock. The read lock is not exclusive --- A range can have multiple locks by different users. 

On the other hand, there is a concept of write lock in the context of FDB per-range lock --- If a user/app takes a writeLock on a range, 
other user/app cannot do any read nor write. The write lock is exclusive, if a user/app takes the writeLock on a range, the range must have not any other lock.
Currently, we only implemented the readLock. The writeLock is currently not implemented. we will implement the writeLock later on demand. 

Example use cases
-----------------
Currently, BulkLoad feature is an example of using the per-range lock. 
BulkLoad is developed to load a large amount of data into a FDB range without going through the transaction system.
As a result, BulkLoad must ensure the correctness during the data injection in the presence of user traffic. 

> To achieve this, BulkLoad locks the range that it is going to write to, and then it writes the data to the range.

Reviewer: ```suggestion
To achieve this, BulkLoad locks the range to prevent user traffic from writing to the database, and then loads the range data directly into storage servers.
```

After the data injection is done, BulkLoad releases the lock.

How to use?
-----------
Currently, FDB only provides ManagementAPI to lock a range. 
Before a user can lock a range, the user must register its identity to the database.
A range can only be locked by a registered owner.
The user can use the following API to register an identity and lock a range.

Lock a range

``ACTOR Future<Void> takeReadLockOnRange(Database cx, KeyRange range, std::string ownerUniqueID);``

Unlock a user range

``ACTOR Future<Void> releaseReadLockOnRange(Database cx, KeyRange range, std::string ownerUniqueID);``

Get locked ranges within the input range

``ACTOR Future<std::vector<KeyRange>> getReadLockOnRange(Database cx, KeyRange range);``

Register a rangeLock owner to database metadata.

``ACTOR Future<Void> registerRangeLockOwner(Database cx, std::string uniqueId, std::string description);``

Remove an owner from the database metadata

``ACTOR Future<Void> removeRangeLockOwner(Database cx, std::string uniqueId);``

> Get all registered rangeLock owner

Reviewer: ```suggestion
Get all registered rangeLock owners
```

``ACTOR Future<std::vector<RangeLockOwner>> getAllRangeLockOwners(Database cx);``

Get the owner of a rangeLock

``ACTOR Future<Optional<RangeLockOwner>> getRangeLockOwner(Database cx, std::string uniqueId);``

Example usage
-------------
Turn on/off user traffic for bulk load based on range lock. Note that ideally, we want to use the write lock to achieve this, now we use the read lock as a temporarily solution.

``ACTOR Future<Void> turnOffUserWriteTrafficForBulkLoad(Transaction* tr, KeyRange range);``

``ACTOR Future<Void> turnOnUserWriteTrafficForBulkLoad(Transaction* tr, KeyRange range);``

Per-Range Lock Design
=====================
The per-range lock information is persisted in ``\xff/rangeLock/`` system key space.
Users specify ranges to lock in ``\xff/rangeLock/`` system key space via a transaction. 
The range lock can be only within user key space, aka ``"" ~ \xff``.

> The value of the key space is either empty or setting as a set of locks. 

Reviewer: ```suggestion
The value within the key space is either empty or a set of locks.
```

> If a range having a ReadLock, the cluster will reject any mutation within the range. 

Reviewer: ```suggestion
If a range has a ReadLock, the cluster will block any transaction that writes to the range. 
```

If the range value is empty, the cluster does not reject the traffic as the range is unlocked.

The per-range lock API issues transactions to update the ``\xff/rangeLock/`` system metadata. 

> When the mutation arrives at commit proxies, the commit proxies keep the in-memory key range map of range locking status, 

Reviewer: ```suggestion
When the mutation arrives at commit proxies, each commit proxy keeps an in-memory key range map of range locking status, 
```

> and they update the map when applying the metadata mutation. 

Reviewer: ```suggestion
and they update the maps when applying the metadata mutation. 
```

When following normal mutation arrives at the commit proxy, the commit proxy checks the range locking status in the in-memory map.

> If the mutation is within a locked range, the commit proxy rejects the mutation.

Reviewer: ```suggestion
If the mutation is within a locked range, the commit proxy rejects the transaction containing the mutation.
```

This process happens in the phase of postResolution, where the commit proxy scans all transactions. 
In addition, the lockRange mutation is also persisted to txnStateStore. When recovery, a new commit proxy will rebuild the in-memory lockRange 
map based on the status persisted in txnStateStore.

When a locking mutation arrives at a commit proxy, 
the lock takes effect on all mutations that arrive at any commit proxy after this locking mutation. 
Note that it is possible that this locking mutation is batched with other mutations with the same commit version. 
The locking mutation takes effect on all mutations after the locking mutation in the same batch.

> To achieve this, the locking transaction adds a write_conflict_range on the entire user key space (``"" ~ \xff``).

Reviewer: is this correct? I thought the write conflict range is set to be the locked range.

Author: Nice catch. I think you are correct. We only need to add the locked range.

Reviewer: OK. Can you update the doc here?

> As a result, any following mutation to the locked range will be marked as ``Conflict``. 

Reviewer: ```suggestion
As a result, any following transactions in the batch that writes to the locked range will be marked as ``Conflict``. 
```

Support multiple range lock users
---------------------------------
To support rangeLock for multiple applications, we add ownership concept to rangeLock. Any range can have multiple locks with different owners. 

> An owner can only unlock its own rangeLock. A mutation will be reject if it goes to a range that has a rangeLock with an owner. 

Reviewer: ```suggestion
An owner can only unlock its own rangeLock. A mutation will be rejected if it updates keys in a range that has a rangeLock with an owner. 
```
